Refactored Linting & Unit Testing Workflows

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This pull request consolidates all of the linting and unit testing workflows in order to make maintenance easier. Having a single workflow allows us to have one central point to optimize as well as provide support to new projects without having to create new workflows. As part of this pull request I am making it so the tasks will only run on projects affected by the pull request. This will improve performance and avoid running unnecessary tasks.

Closes #40285, #40286.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the new workflow run in the PR.
2. Review workflows documented in the comments.
3. Create a draft PR from this branch targeting it.
4. Try PHP-only changes
5. Try JS-only changes
6. Try no changes.
7. Try Markdown only changes
8. Try test file only changes (for JS and PHP respectively)
9. Compare generated workflows to existing workflows. This includes WooCommerce Core's testing suite, the other plugin testing suite, and linting workflows. All of the workflows deleted in this PR, basically.

One thing not covered in the testing instructions above is whether or not the workflow works post-merge. I did most of my testing around that locally by looking at the output of the `build-ci-matrix.js` file. When you call `buildCIMatrix()` with a falsey value it adds every project and every type of command that it has.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

Just making some changes to support the new CI workflow.

</details>

ObliviousHarmony committed:
Remove Prime Cache Workflow

This workflow isn't necessary because the normal CI workflows will
populate the cache when they run on `trunk`.

ObliviousHarmony committed:
Separated PHP and JS Unit Tests

This identifies test suites using a `test:js` or `test:php` script. To
support existing workflows the `test` script now calls the new one.

ObliviousHarmony committed:
Created New Unit Test Workflow

This new workflow will replace the myriad of unit testing
workflows that are currently present within the monorepo.
It only runs on affected projects and supports automatically
recognizing new unit tests for new projects.

ObliviousHarmony committed:
Limited Push Event Scope

This avoids us running the action twice on pull requests.

ObliviousHarmony committed:
Added Concurrency Cancellation

Let's not run the workflow more than once at a time!

ObliviousHarmony committed:
Fixed Workflow OS

ObliviousHarmony committed:
Fixed Matrix Branch Comparison

ObliviousHarmony committed:
Added Monorepo Setup

Long-term this action should be broken up to remove any
unnecessary steps, but for now, it's needed to easily
consolidate the usage of PNPM.

ObliviousHarmony committed:
Cleaned Up Job Dependencies

ObliviousHarmony committed:
Speed Up `pnpm install`

This avoids the Chromium download that takes a lot time.

ObliviousHarmony committed:
Fixed Indentation

ObliviousHarmony committed:
Fixed Comparison Ref

ObliviousHarmony committed:
Improved Monorepo Setup Performance

We now add support for _not_ installing dependencies.

ObliviousHarmony committed:
Use Branch SHA

ObliviousHarmony committed:
Changed Job Dependencies

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Base Ref

ObliviousHarmony committed:
Trying Ref Fix

ObliviousHarmony committed:
Fixed Base Ref

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Increased Fetch Depth

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Matrices Construction

ObliviousHarmony committed:
Matrix Test

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Job Output

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fix Matrix

ObliviousHarmony committed:
Cleaned Up Matrix Construction

Now that the construction is working, let's go ahead and
clean up the matrix construction and make sure everything
works the way that it's supposed to.

ObliviousHarmony committed:
Trying ENV Parsing

ObliviousHarmony committed:
Removed Unnecessary Cache Priming

ObliviousHarmony committed:
Simplified Matrix Generation

ObliviousHarmony committed:
Removed Replaced Workflows

Since we've consolidated these workflows we have no
reason to keep the workflow files around anymore.

ObliviousHarmony committed:
Centralized Matrices Generation

Since we're going to consolidate _all_ of the workflows,
I've gone ahead and created a single file to use in
generating the CI matrices. For now this only generates
matrices for testing but it's immediately ready for more.

ObliviousHarmony committed:
Added Additional Task Support

This adds support for defining additional tasks to run
using the project's `package.json` file.

ObliviousHarmony committed:
Updated Workflow

This updates the workflow to make use of the new matrix
generation method.

ObliviousHarmony committed:
Fixed Expected `buildCIMatrix` Output

Previously we were using a promise and now we don't
have to.

ObliviousHarmony committed:
Fixed Command

ObliviousHarmony committed:
Separated JS and PHP Tests

ObliviousHarmony committed:
Fix Monorepo Install Filter

ObliviousHarmony committed:
Added Lint & E2E Support

ObliviousHarmony committed:
Fixed Variable Types

ObliviousHarmony committed:
Test Change

This _should_ trigger a dependency JS build & test but not
a PHP one.

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/unit-test-workflow

ObliviousHarmony committed:
Skip Unnecessary Build Caching

ObliviousHarmony committed:
Fixed Command Possibility Evaluation

Rather than requiring all of the criteria to match we
want to consider a single matching criteria a pass.

ObliviousHarmony committed:
Replaced Lint Workflows

Since this workflow now replaces testing and linting we are
able to remove the other workflows.

ObliviousHarmony committed:
Fixed File Path

ObliviousHarmony committed:
Added WP Version Parsing

For handling the test environment we allow a WP
version that gets converted to a download URL.

ObliviousHarmony committed:
Added WooCommerce WP Version Matrices

ObliviousHarmony committed:
Added Environment Var Parsing

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Documentation

ObliviousHarmony committed:
Testing Env Unwrapping

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Removed Logging

Looks like it works!

ObliviousHarmony committed:
Run Tasks With Errors

Even when a failure for a specific step happens we should
still run the others. This changes the steps so that they
will stull run even if a previous one fails.

ObliviousHarmony committed:
Changed Status Function Order

ObliviousHarmony committed:
Added Post-Matrix Status Check Job

We can set this job as a required status check and it will only
pass if all of the projects in the matrix pass as well.

ObliviousHarmony committed:
Test PHP Change

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Additional Task Commands

ObliviousHarmony committed:
Fixed Unit Test Test Check

ObliviousHarmony committed:
Revert "Separated PHP and JS Unit Tests"

This reverts commit efef597bd511e70ca1a5e820c52e5c6ff60adf5e.

ObliviousHarmony committed:
Added Custom Command Override Support

In cases where we might not be able to standardize a
command for use in CI we should be able to support
using a custom command in place of the original.

ObliviousHarmony committed:
Added Token Replacement

This will let us replace tokens in a command so that we can
go a step further with command customization. This is
pretty much just in support of running phpcs-changed.

ObliviousHarmony committed:
Linting Fixes

ObliviousHarmony committed:
Fixed Unnecessary Task Additions

The additional tasks weren't being ignored and checked against
the possible commands.

ObliviousHarmony committed:
Added Custom Command To Possible Command Check

In cases where the default command doesn't exist we were
getting false negatives on whether a command can run. This
uses the custom command in validation where possible to
ensure that it gets checked correctly.

ObliviousHarmony committed:
Linting Fixes

ObliviousHarmony committed:
Added Custom Command Validation

We need to validate the custom commands and make
sure that the script is callable.

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/unit-test-workflow

ObliviousHarmony committed:
Added Custom Test Env Command

ObliviousHarmony committed:
Refactored Config File Validation

This takes an extra step to make sure we're loading
valid config options.

ObliviousHarmony committed:
Added WP Version URL Support

Instead of always requiring the user to pass a version we
can also support them just putting a URL in themselves.

ObliviousHarmony committed:
Use Command Enum

Rather than using the raw command strings we're going to
switch to an enum. This will make it much easier to validate
and reason about the specific commands available.

ObliviousHarmony committed:
Added Command Type Validation

This cleans up the validation for command configuration to
make sure there aren't any accidental configurations.

ObliviousHarmony committed:
Disabled E2E Test Command

This one will require some thought about whether it even
belongs here. E2E tests are much more complicated.

ObliviousHarmony committed:
Removed Workflow Matrix Debug

ObliviousHarmony committed:
Documented CI Workflow

ObliviousHarmony committed:
Removed PHP Linting Workflow

We don't need it anymore! I also removed an unnecessary script.

ObliviousHarmony committed:
Removed Install Filter For Debugging

One of the builds are always failing and I think it might
have something to do with messed up transitive
dependencies. This will help highlight if that's true.

ObliviousHarmony committed:
Removed E2E From CI Workflow

This probably won't be a good home. I'm going to leave the
detection handling in the matrix because it _might_ make
sense to have a separate E2E matrix to draw from using
the same change detection data. A separate kind of job?

ObliviousHarmony committed:
Cleaned Up Validation

I moved all of the validation into a single function to make
it easier to understand and test.

ObliviousHarmony committed:
Fixed Token Replacement

ObliviousHarmony committed:
Expanded Checkout Depth

ObliviousHarmony committed:
Fixed Custom Command Error

ObliviousHarmony committed:
Fixed Test Env Creation Check

ObliviousHarmony committed:
Fixed Test ENV Command Validation

ObliviousHarmony committed:
Remove PHP & JS Modifications

ObliviousHarmony committed:
Fixed Accidental Test Config Mistake

ObliviousHarmony commented:
Here are two runs of this workflow visible here in this PR:

* JS & PHP Changes: https://github.com/woocommerce/woocommerce/actions/runs/6461108496?pr=40572
  * I introduced a PHPCS linting error here to show that the `phpcs-changed` works.
* WooCommerce Core JS Change Only: https://github.com/woocommerce/woocommerce/actions/runs/6462414863?pr=40572
  * This necessarily exists because we modified the `package.json` file for Core.

ObliviousHarmony commented:
In the future it might make sense to refactor the `testEnv` configuration to support per-command test environments. I can see cases where different types of tests may have different requirements. This seems like it'll probably be necessary when we consolidate the E2E tests. For now though, I think this is ready for review.

ObliviousHarmony committed:
Fixed testEnvCommand Filtering Check

ObliviousHarmony committed:
Made Linting Script Shell Agnostic

ObliviousHarmony committed:
Attempting Shell Fix

ObliviousHarmony committed:
Adjusting Shebang

ObliviousHarmony committed:
Trying Shell Change

ObliviousHarmony committed:
Debugging

ObliviousHarmony committed:
Testing

ObliviousHarmony committed:
Fixed File Path Regex

ObliviousHarmony committed:
Fixed Broken Regecx

ObliviousHarmony committed:
Fixed Change Detection Resetting

ObliviousHarmony committed:
Fixed Script

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Fixed Matrix Evaluation

invalid-email-address committed:
Add changefile(s) from automation for the following project(s): woocommerce

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Merge remote-tracking branch 'refs/remotes/origin/refactor/unit-test-workflow' into refactor/unit-test-workflow

ObliviousHarmony committed:
Merge remote-tracking branch 'refs/remotes/origin/refactor/unit-test-workflow' into refactor/unit-test-workflow

ObliviousHarmony committed:
Fixed Empty Matrices

ObliviousHarmony committed:
Fix Matrix Workflow Error

ObliviousHarmony committed:
Fixed Workflow Syntax

ObliviousHarmony committed:
Add Nicer Empty Matrix Messaging

ObliviousHarmony committed:
Attempt #2

ObliviousHarmony committed:
Name Expression Aren't Evaluated

ObliviousHarmony commented:
Test Cases from #40674:

* Documentation Change: https://github.com/woocommerce/woocommerce/actions/runs/6473823011?pr=40674
  * There's currently a workflow that lints markdown but I'm not sure if that linting belongs in this workflow. It makes sense but there is some top-level documentation too. Maybe we need to consider changes outside of a project for the workflow? A problem for tomorrow.
* JS Test File Change: https://github.com/woocommerce/woocommerce/actions/runs/6473886770?pr=40674

psealock commented on .github/workflows/ci.yml:
```diff
@@ -1,90 +1,98 @@
+                baseRef = 'origin/' + baseRef;
+              }
+              const buildCIMatrix = require( './.github/workflows/scripts/build-ci-matrix' );
+              core.setOutput( 'matrix', JSON.stringify( await buildCIMatrix( baseRef ) ) );
```
Some logging would be useful here.

I introduced [a lint error in `@woocommerce/components`](https://github.com/woocommerce/woocommerce/pull/40691/commits/79422d1c070639564d199b50134ab2904f8b0775) and I do get a failure in JS tests for `woocommerce/client/admin`. I'm trying to debug and see if that's even related. `@woocommerce/components` tests weren't run. See [PR here](https://github.com/woocommerce/woocommerce/pull/40691)

psealock replied:
Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

psealock replied:
Upon restarting, I do get the components to lint, but it passes 🤷🏽 https://github.com/woocommerce/woocommerce/actions/runs/6476659736/job/17588896320?pr=40691

ObliviousHarmony replied:
> Some logging would be useful here.

What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like. The only information that I can think of that might be "hidden" is the intermediary steps leading up to the creation of the matrix. Namely, the files that were changed (and the projects they marked as changed) and the changes that were marked (along with the file that marked them)? This seems like it _could_ be interesting for being able to track how a project got added to the matrix. I'll take a look at adding that.

> Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

I'm not really sure what value we could extract from this step. All it does is check for failed jobs so the error indicating that a job has failed is just telling you what you can already see from GitHub's interface on the job list. The `needs` context doesn't provide us information about _which_ jobs failed, but again, you can actually just see that. Once you click on the job you'll see what steps failed.

psealock replied:
>What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like

I made a change in a particular file which should kick off certain checks, which didn't fire. My attempts to figure out what went wrong (at the matrix creation step? At the matrix interpretation step?) would have been aided by some logging. For instance, a print out of the matrix itself would have let me know that indeed the changes made were correctly picked up (or not) and given me a clue as to where the problem might be. 

I find it useful for tools like this to be overly descriptive. "Here's what I'm dong", "Here's what the result is", and "Here is what I'm going to do next".

ObliviousHarmony replied:
I added some pretty extensive logging, take a look and let me know what you think.

ObliviousHarmony committed:
Cleaned Up File Change Detection

ObliviousHarmony committed:
Added Logging

This provides some insight into how the script works so that
debugging is easier.

ObliviousHarmony commented:
I've pushed another commit that adds logging and more granular detection of asset file changes. This _might_ have created problems with change detection though so that'll need looked at in testing.

ObliviousHarmony committed:
Expanded Logging

ObliviousHarmony committed:
Skip Empty Path

jonathansadowski reviewed (approved):
Sorry it took me so long to finish the review of this, I had some false-starts last week 😄 

I've compared the output of the workflow here to the output of the workflows with similar PRs against trunk, and for the parts of the workflow that run, I'm not noticing any difference... and unnecessary tests are not running (no PHP tests for JS changes, and vise-versa).

I've also walked through the changes, as well as the code that generates the matrix, and everything looks good to me, nice work!

I'm approving this, but given its scope, I'd be more comfortable if at least somebody else also approved, but I'll leave that up to you.

psealock commented:
Thanks for adding the logging, it makes it really transparent whats happening.

For this change in https://github.com/woocommerce/woocommerce/pull/40857, I see `test:js` getting added to the matrix. Perhaps because its detecting a change in `package.json` and `composer.json`? 

<img width="789" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/53b85f6a-b20a-4d9b-b6e9-673fadf0ef4d">

<img width="938" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/842b9d15-202c-4c0c-945b-b8b134618b1a">

ObliviousHarmony commented:
Changes to any JSON file are registered as JS source changes @psealock. This catches the MANY cases of JSON config file changes that can impact linting and testing for JS projects. In the future we can get more granular with detection but the heavy-handed approach is at-worst no different than the current one and at-best an improvement when it works

psealock commented:
JS only change in a component nicely kicks off JS tests in all affected packages and plugins https://github.com/woocommerce/woocommerce/pull/40859

<img width="794" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/1dcf9562-ca90-4f39-912a-031f2d887d75">


<img width="812" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/a60fcfa0-de74-4ad7-a64d-1d4a972cd203">

Interestingly [CI passes](https://github.com/woocommerce/woocommerce/actions/runs/6567854388/job/17841330136?pr=40859), but this passes locally as well - although it shouldn't given the change. This might be something unrelated.

```
pnpm --filter="@woocommerce/components" lint
```

psealock commented:
>In the future we can get more granular with detection but the heavy-handed approach is at-worst no different than the current one and at-best an improvement when it works

Cool cool, this makes sense. I changed the diff in the same file and got different results, which seem more in line with expectations.

<img width="768" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/f42cd7e1-551a-409d-b68f-ff35d1f05402">

<img width="821" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/b9f9dc56-66b5-4b2a-ab7b-b3114ea110d2">

psealock reviewed (approved):
I've done a fair bit of testing and since any failures anywhere can be gathered post-merge, we should be able to know if anything falls through the cracks.

Nice job with this @ObliviousHarmony !
